// @tarojs/plugin-platform-swan/dist/index.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var service = require('@tarojs/service');
var shared = require('@tarojs/shared');
var template = require('@tarojs/shared/dist/template');

const swanSpecialAttrs = {
    'scroll-view': ['scrollTop', 'scrollLeft', 'scrollIntoView'],
    'movable-view': ['x', 'y'],
    slider: ['value'],
    input: ['value'],
    textarea: ['value']
};
class Template extends template.RecursiveTemplate {
    constructor(options) {
        var _a, _b;
        super();
        this.supportXS = true;
        this.Adapter = {
            if: 's-if',
            else: 's-else',
            elseif: 's-elif',
            for: 's-for',
            forItem: 's-for-item',
            forIndex: 's-for-index',
            key: 's-key',
            xs: 'sjs',
            type: 'swan'
        };
        this.buildFlattenView = (level = this.flattenViewLevel) => {
            if (level === 0) {
                return '<template is="{{xs.e(0)}}" data="{{{i:item}}}" />';
            }
            const child = this.buildFlattenView(level - 1);
            const template = `<view s-if="{{item.nn==='view'&&(item.st||item.cl)}}" id="{{item.uid}}" ${this.buildFlattenNodeAttributes('view')}>
  <block s-for="{{item.cn}}" s-key="uid">
    ${shared.indent(child, 4)}
  </block>
</view>
<text s-elif="{{item.nn==='text'&&(item.st||item.cl)}}" id="{{item.uid}}" ${this.buildFlattenNodeAttributes('text')}>
  <block s-for="{{item.cn}}" s-key="uid">
    <block>{{item.v}}</block>
  </block>
</text>
<text s-elif="{{item.nn==='static-text'&&(item.st||item.cl)}}" id="{{item.uid}}" ${this.buildFlattenNodeAttributes('static-text')}>
  <block s-for="{{item.cn}}" s-key="uid">
    <block>{{item.v}}</block>
  </block>
</text>
<button s-elif="{{item.nn==='button'&&(item.st||item.cl)}}" id="{{item.uid}}" ${this.buildFlattenNodeAttributes('button')}>
  <block s-for="{{item.cn}}" s-key="uid">
    <template is="{{xs.e(0)}}" data="{{{ i:item }}}" />
  </block>
</button>
<input s-elif="{{item.nn==='input'&&(item.st||item.cl)}}" id="{{item.uid}}" ${this.buildFlattenNodeAttributes('input')} />
<swiper s-elif="{{item.nn==='swiper'&&(item.st||item.cl)}}" id="{{item.uid}}" ${this.buildFlattenNodeAttributes('swiper')}>
  <block s-for="{{item.cn}}" s-key="uid">
    <template is="{{xs.e(0)}}" data="{{{ i:item }}}" />
  </block>
</swiper>
<block s-else>
  <template is="{{xs.e(0)}}" data="{{{i:item}}}" />
</block>`;
            return template;
        };
        this.buildFlattenCover = (level = this.flattenCoverLevel) => {
            if (level === 0) {
                return '<template is="{{xs.e(0)}}" data="{{{i:item}}}" />';
            }
            const child = this.buildFlattenCover(level - 1);
            const template = `<cover-view s-if="{{item.nn==='cover-view'}}" id="{{item.uid}}" ${this.buildFlattenNodeAttributes('cover-view')}>
  <block s-for="{{item.cn}}" s-key="uid">
    ${shared.indent(child, 4)}
  </block>
</cover-view>
<cover-image s-elif="{{item.nn==='cover-image'}}" id="{{item.uid}}"  ${this.buildFlattenNodeAttributes('cover-image')}></cover-image>
<block s-else>
  <template is="{{xs.e(0)}}" data="{{{i:item}}}" />
</block>`;
            return template;
        };
        this.modifyLoopBody = (child, nodeName) => {
            switch (nodeName) {
                case 'view':
                    // fix issue #6015
                    return this.buildFlattenView();
                case 'cover-view':
                case 'canvas':
                case 'map':
                case 'animation-view':
                case 'camera':
                case 'live-player':
                    return this.buildFlattenCover();
                case 'video': {
                    const body = `<ad s-if={{item.nn==='ad'}} id="{{item.uid}}" ${this.buildFlattenNodeAttributes('ad')}></ad>
<block s-else>
  ${shared.indent(this.buildFlattenCover(), 2)}
</block>`;
                    return body;
                }
                case 'text':
                case 'static-text':
                    return `<block>{{i.${"cn" /* Childnodes */}[index].${"v" /* Text */}}}</block>`;
                case 'picker-view':
                    return `<picker-view-column id="{{item.uid}}" ${this.buildFlattenNodeAttributes('picker-view-column')}>
          <block s-for="{{item.cn}}" s-key="uid">
            ${child}
          </block>
        </picker-view-column>`;
                default:
                    return child;
            }
        };
        this.modifyLoopContainer = (children, nodeName) => {
            if (nodeName === 'swiper') {
                return children.replace(/s-for="{{i\.cn}}"/, 's-for="{{xs.f(i.cn)}}"');
            }
            return children;
        };
        this.modifyTemplateResult = (res, nodeName) => {
            if (nodeName === 'picker-view-column')
                return '';
            return res;
        };
        this.flattenViewLevel = (_a = options === null || options === void 0 ? void 0 : options.flattenViewLevel) !== null && _a !== void 0 ? _a : 8;
        this.flattenCoverLevel = (_b = options === null || options === void 0 ? void 0 : options.flattenViewLevel) !== null && _b !== void 0 ? _b : 3;
    }
    createMiniComponents(components) {
        const result = super.createMiniComponents(components);
        this.legacyMiniComponents = Object.assign({}, result);
        delete result['pure-view'];
        delete result['static-view'];
        return result;
    }
    buildXsTemplate() {
        return '<import-sjs module="xs" src="./utils.sjs" />';
    }
    dataKeymap(keymap) {
        return `{ ${keymap} }`;
    }
    getAttrValue(value, key, nodeName) {
        if (shared.isArray(swanSpecialAttrs[nodeName]) && swanSpecialAttrs[nodeName].includes(key)) {
            return `= ${value} =`;
        }
        return `{ ${value} }`;
    }
    buildFlattenNodeAttributes(nodeName) {
        const component = this.legacyMiniComponents[nodeName];
        return Object.keys(component)
            .map(k => `${k}="${k.startsWith('bind') || k.startsWith('on') || k.startsWith('catch') ? component[k] : `{{${component[k].replace('i.', 'item.')}}}`}"`)
            .join(' ');
    }
    buildXSTmpExtra() {
        return `f: function (l) {
    return l.filter(function (i) {return i.nn === 'swiper-item'})
  }`;
    }
}

const components = {
    // ======== 调整属性 ========
    Progress: {
        'border-radius': '0',
        'font-size': '16',
        duration: '30'
    },
    RichText: {
        selectable: 'false',
        name: '',
        attrs: '',
        children: '[]',
        text: '',
        'image-menu-prevent': 'false',
        preview: ''
    },
    Map: {
        polygons: '[]',
        'enable-3D': 'false',
        'show-compass': 'false',
        'enable-overlooking': 'false',
        'enable-zoom': 'true',
        'enable-scroll': 'true',
        'enable-rotate': 'false',
        bindRegionChange: '',
        bindPoiTap: ''
    },
    Button: {
        bindGetPhoneNumber: '',
        bindGetUserInfo: '',
        bindOpenSetting: '',
        bindContact: '',
        bindChooseAddress: '',
        bindChooseInvoiceTitle: ''
    },
    Form: {
        'report-type': 'default',
        'template-id': '',
        'subscribe-id': ''
    },
    Input: {
        'adjust-position': 'true'
    },
    Textarea: {
        'confirm-type': shared.singleQuote('default'),
        'confirm-hold': 'false',
        'show-confirm-bar': 'true',
        'adjust-position': 'true'
    },
    Navigator: {
        target: shared.singleQuote('self'),
        'app-id': '',
        path: '',
        'extra-data': '',
        version: shared.singleQuote('version')
    },
    Image: {
        webp: 'false',
        'image-menu-prevent': 'false',
        preview: '',
        'original-src': ''
    },
    Video: {
        title: '',
        'show-no-wifi-tip': 'true',
        'vslide-gesture': 'false',
        'vslide-gesture-in-fullscreen': 'true',
        'enable-play-gesture': 'false',
        'show-rate-btn': 'false',
        'show-vslide-btn-in-fullscreen': 'true',
        'silent-play': 'false',
        bindLoadedMetadata: ''
    },
    Ad: {
        appid: '',
        apid: '',
        type: shared.singleQuote('feed'),
        updatetime: '',
        bindStatus: ''
    },
    // ======== 额外组件 ========
    Tabs: {
        'tabs-background-color': shared.singleQuote('#fff'),
        'tabs-active-text-color': shared.singleQuote('#000'),
        'tabs-inactive-text-color': shared.singleQuote('#666'),
        'tabs-underline-color': shared.singleQuote('#333'),
        'active-name': '',
        'url-query-name': '',
        'max-tab-item-amount': '5',
        bindTabChange: ''
    },
    TabItem: {
        label: '',
        name: '',
        'badge-type': '',
        'badge-text': ''
    },
    AnimationVideo: {
        'resource-width': '800',
        'resource-height': '400',
        'canvas-style': shared.singleQuote('width:400px;height:400px'),
        path: '',
        loop: 'fasle',
        autoplay: 'fasle',
        bindStarted: '',
        bindEnded: ''
    },
    AnimationView: {
        path: '',
        loop: 'false',
        autoplay: 'true',
        action: shared.singleQuote('play'),
        hidden: 'true',
        bindEnded: ''
    },
    ArCamera: {
        key: '',
        type: '',
        flash: shared.singleQuote('off'),
        bindError: '',
        bindLoad: '',
        bindMessage: '',
        bindScanCode: ''
    },
    RtcRoom: {
        id: '',
        'enable-camera': 'true',
        'enable-auto-focus': 'true',
        'enable-zoom': 'false',
        'device-position': shared.singleQuote('front'),
        'enable-mic': 'true',
        'enable-agc': 'false',
        'enable-ans': 'false',
        bitrate: '900',
        'video-width': '360',
        'video-height': '640',
        'enable-remote-mirror': 'false',
        'local-mirror': shared.singleQuote('auto'),
        'sound-mode': shared.singleQuote('speaker'),
        bindStateChange: '',
        bindError: ''
    },
    RtcRoomItem: {
        id: '',
        type: '',
        'user-id': ''
    },
    OpenData: {
        type: ''
    }
};

const PACKAGE_NAME = '@tarojs/plugin-platform-swan';
const PROJECT_JSON = 'project.swan.json';
class Swan extends service.TaroPlatformBase {
    /**
     * 1. setupTransaction - init
     * 2. setup
     * 3. setupTransaction - close
     * 4. buildTransaction - init
     * 5. build
     * 6. buildTransaction - close
     */
    constructor(ctx, config) {
        super(ctx, config);
        this.platform = 'swan';
        this.globalObject = 'swan';
        this.runtimePath = `${PACKAGE_NAME}/dist/runtime`;
        this.taroComponentsPath = `${PACKAGE_NAME}/dist/components-react`;
        this.fileType = {
            templ: '.swan',
            style: '.css',
            config: '.json',
            script: '.js',
            xs: '.sjs'
        };
        this.setupTransaction.addWrapper({
            close() {
                this.modifyComponents();
                ctx.generateFrameworkInfo();
                this.generateProjectConfig(PROJECT_JSON, PROJECT_JSON);
            }
        });
        this.template = new Template({
            flattenViewLevel: config.flattenViewLevel
        });
    }
    /**
     * 增加组件或修改组件属性
     */
    modifyComponents() {
        this.template.mergeComponents(this.ctx, components);
        delete this.template.internalComponents.Input.cursor;
        delete this.template.internalComponents.Input['selection-start'];
        delete this.template.internalComponents.Input['selection-end'];
    }
}

var index = (ctx) => {
    ctx.registerPlatform({
        name: 'swan',
        useConfigName: 'mini',
        async fn({ config }) {
            const program = new Swan(ctx, config);
            await program.start();
        }
    });
};

exports.Swan = Swan;
exports.default = index;
